### PR TITLE
Build guest method crates automatically

### PR DESCRIPTION
**Pull request by shkoo**
* Deprecate risc0-build-methods as it's no longer necessary
* Only recalculate method IDs when necessary since it's slow
* Out-of-tree builds with --target-dir now work

### CODE REVIEW

File: risc0/build/src/lib.rs [hunk 1]
```diff
@@ -82,99 +197,67 @@ pub fn build(manifest_path: &Path, target_dir: &Path) {
...
-        let manifest_path = Path::new(&pkg.manifest_path)
-            .parent()
-            .unwrap()
-            .join(rel_path)
-            .join("Cargo.toml");
+        // Try to rebuild guest packages if anything changes in the guest package directory.
+        println!(
+            "cargo:rerun-if-changed={}",
+            guest_pkg.manifest_path.parent().unwrap()
+        );
```
Comment (flaub) [inline risc0/build/src/lib.rs:250]: Ah yes! This is the secret sauce! Nice work.

Comment (flaub) [reply risc0/build/src/lib.rs:250]: Is this transitive?

**Comment by spaugh** [comment]
holy shit. 😍 😍 😍 😍 😍 😍 😍 😍 😍 😍 😍 😍 😍 😍 😍

File: risc0/build/src/lib.rs [hunk 2]
```diff
@@ -82,99 +197,67 @@ pub fn build(manifest_path: &Path, target_dir: &Path) {
...
         std::process::exit(status.code().unwrap());
     }
 }
 
 /// Embeds methods built for RISC-V for use by host-side dependencies.
 ///
+/// This method should be called from a package with a
+/// [package.metadata.risc0] section including a "methods" property
+/// listing the subdirectories that contain riscv guest method
+/// packages.
```
Comment (flaub) [inline risc0/build/src/lib.rs:216]:
```suggestion
/// This method should be called from a package with a
/// [package.metadata.risc0] section including a "methods" property
/// listing the relative paths that contain riscv guest method
/// packages.
```

File: risc0/build/src/lib.rs [hunk 3]
```diff
@@ -12,171 +12,252 @@
...
         }
     }
+
+    // HACK: It's not particularly practical to figure out all the
+    // files that all the guest crates transtively depend on.  So, we
+    // want to run the guest "cargo build" command each time we build.
+    //
+    // Since we generate methods.rs each time we run, it will always
+    // be changed.
+    println!("cargo:rerun-if-changed={}", methods_path.display());
```
Comment (flaub) [inline risc0/build/src/lib.rs:260]: Love it, this is a super clever hack!

File: risc0/build/src/lib.rs [hunk 4]
```diff
@@ -12,171 +12,252 @@
...
     process::Command,
 };
 
-use cargo_metadata::MetadataCommand;
+use cargo_metadata::{MetadataCommand, Package};
 use risc0_zkvm_platform_sys::LINKER_SCRIPT;
 use risc0_zkvm_sys::MethodID;
 use serde::Deserialize;
-
-const TARGET_JSON: &[u8] = include_bytes!("../riscv32im-unknown-none-elf.json");
```
Comment (flaub) [inline risc0/build/src/lib.rs:30]: Let's remove this file from the repo now that it's not needed

File: risc0/build/src/lib.rs [hunk 5]
```diff
@@ -12,171 +12,252 @@
...
+{
+    fs::create_dir_all(target_dir.as_ref()).unwrap();
+
+    let cargo = env::var("CARGO").unwrap();
     let args = vec![
         "build",
         "--release",
         "--target",
-        target_path.to_str().unwrap(),
+	"riscv32im-unknown-none-elf",
```
Comment (flaub) [inline risc0/build/src/lib.rs:192]: This seems like a rustfmt issue?

**Review by flaub — APPROVED** [review]

**Comment by flaub** [comment]
@shkoo Does this PR now fix #116?

**Pull request merged**